File: companies/Tesla/229.majority-element-ii.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        /*
        Approach: Boyer-Moore generalized for elements appearing more than n/3.
        There can be at most two such elements, so maintain two candidates and
        counters, then verify them with a second pass.

        Complexity: O(n) time, O(1) space.
        */
        int cand1 = 0, cand2 = 1, count1 = 0, count2 = 0;
        for (int num : nums) {
            if (num == cand1) ++count1;
            else if (num == cand2) ++count2;
            else if (count1 == 0) { cand1 = num; count1 = 1; }
            else if (count2 == 0) { cand2 = num; count2 = 1; }
            else { --count1; --count2; }
        }
        vector<int> ans;
        count1 = count2 = 0;
        for (int num : nums) {
            if (num == cand1) ++count1;
            else if (num == cand2) ++count2;
        }
        if (count1 > (int)nums.size() / 3) ans.push_back(cand1);
        if (count2 > (int)nums.size() / 3) ans.push_back(cand2);
        return ans;
    }
};

File: companies/Amazon/229.majority-element-ii.py
"""
Approach: Boyer-Moore voting generalized for elements appearing more than n/3 times.
Data structure: at most two candidates and their counts are needed because there can be no more than two values with frequency greater than n/3.
Interview logic: pairs of different non-candidate elements can be canceled against the two candidates. A second pass verifies the surviving candidates because the first pass only finds possibilities.
Complexity: O(n) time, O(1) space.
Tests and edge cases: empty list returns []; one or two elements; candidate duplicates are filtered by verification.
"""
from __future__ import annotations
from typing import List

# @lc code=start
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        cand1 = cand2 = None
        count1 = count2 = 0
        for num in nums:
            if num == cand1:
                count1 += 1
            elif num == cand2:
                count2 += 1
            elif count1 == 0:
                cand1, count1 = num, 1
            elif count2 == 0:
                cand2, count2 = num, 1
            else:
                count1 -= 1
                count2 -= 1
        return [cand for cand in (cand1, cand2) if cand is not None and nums.count(cand) > len(nums) // 3]
# @lc code=end
